Queryable Unified Inventory API

Mumfrey opened the pull request:
This pull request encapsulates a proposal for an Inventory API for Sponge. At the core it is based upon #242 by @gratimax but is revised and expanded to include the proposition outlined below. Please read all of this document before commenting on design specifics within this PR.

This is my first PR to the API side of things so be gentle.

Design Proposal for a Queryable Unified Inventory API
----------------

Many of the proposed systems for dealing with inventories in Sponge have been very close to the underlying Minecraft implementation of inventories, there are a number of drawbacks with this model:

* Inventory topology is bifurcated into "Inventories", generally describing an actual container (the *Model*), and "Containers" which present a code-facing view of one or more "Inventories" (the *ViewModel*). Which has its own problems:
 * Generally the physical layout of the Inventory and the physical layout of the Container are not contractually linked in any meaningful way.
 * The terminology when referring to one or other of these constructs gets very wooly very quickly (especially with Containers that have a 1:1 mapping of "slots" to underlying Inventory position).
 
 The *View* in this instance (taking MVVM as our template) is a client-only user-facing object responsible for displaying the Container to the user.
 
* Working with different types of inventory requires prior knowledge of the inventory's internal layout (interface is coupled invisibly to implementation) because there is no way to obtain *meta* information about a specific Inventory or Container. This means that:
 * consumers have to deal with the characteristics of specific inventories, and any code which wants to deal with different types of inventory must do so by handling each inventory type directly
 * This makes the API extremely brittle and not particularly extensible, since any types of inventory which are not *known about in advance* cannot be efficiently interacted with
 * Code which consumes the inventory API is *forced* to make assumptions which must then be forcibly invalidated (for example by throwing an exception) should the underlying behaviour of a particular inventory change. eg. the same brittleness applies to "built in" inventories as to mod-provided inventories over time and there is no way to deal gracefully with this.
 * In other words, assumptions *have* to be made, but there is no programmatic way for a consumer to *validate* those assumptions prior to operation
* Because of the blurring of the distinction between Inventory and Container, functionality is cross-contaminated between the two concepts, and this lack of separation of concerns tends to (in turn) pollute any API which tries to faithfully replicate it.

### Design Goals

The aim of this proposal is to address these shortcomings in the following ways:

* Devise an API where *describing* an Inventory is a core aspect of interacting with an Inventory, essentially making Inventories *aspect-oriented*.
* Provision for extensibility by allowing consumers to *query* an Inventory for the particular *aspect* they are interested in, and manipulate the inventory via the results of their queries.
* Create a clear separation between Inventory and Container concepts but acknowledge their overlaps, by making Container essentially another aspect that an Inventory can have.

### Starting Assumptions

As Morpheus famously said to Neo: *"free your mind"*.

To begin talking about a new structure for Inventory interaction, it's necessary to forget everything you currently understand about Inventories. Done that? Okay, let's start with the basic concepts.

#### Common Inventory Operations

If we make **no other assumptions** about an inventory, we can essentially begin by declaring that, in many ways an Inventory exhibits the same behaviour as a `Queue` in that we can add items to the Inventory, consume items from the Inventory, and get the underlying capacity of the Inventory.

```java
interface Inventory {
	/**
	 * Get and remove the first available stack from this
	 * Inventory
	 */
	public abstract Optional<ItemStack> poll();
	
	/**
	 * Get without removing the first available stack from this
	 * Inventory
	 */
	public abstract Optional<ItemStack> peek();
	
	/**
	 * Try to put an ItemStack into this Inventory. Just like
	 * Queue, this method returns true if the Inventory accepted
	 * the stack and false if not, the size of the supplied
	 * stack is reduced by the number of items successfully
	 * consumed by the Inventory.
	 */
	public abstract boolean offer(ItemStack stack);
	
	/**
	 * The number of stacks in the Inventory
	 */
	public abstract int size();
	
	/**
	 * The maximum number of stacks the Inventory can hold
	 */
	public abstract int capacity();
}
```

Note that our base implementations of `poll` and `offer` don't provide any position information, this is because we are not making any other assumptions about the structure of the Inventory at this point. While this might seem like an odd baseline, it begins to make sense when we consider that with a query-based Inventory API, we will be returning inventories which fall into one of the following three categories:

* **Empty Inventory** - to provide for a fluent interface we don't want to return null, the **Empty Inventory** will be our representation of an *empty result set*.
* **Single-Stack Inventory** - an `Inventory` with a single stack is effectively a `Slot`. By keeping the `Inventory` interface simple enough that a `Slot` can reasonably extend it, we have simple unified way of a query returning anything from an entire Inventory, to a specific set of matching slots, to a single slot containing an Item we're looking for.
* **Multi-Stack Inventory** - other inventories which can contain more than a single stack.

#### Basis and Purpose of Queries

Before we look at how queries can work, let's first take a look at a simple use-case to understand the purpose of queries in the first place.

The `PlayerInventory` class actually encompasses several groups of inventory slots we may be interested in. These groups are internally (physically) separated into two arrays: the **Main Inventory** and the **Armour Inventory**. The **Main Inventory** comprises both the hotbar area (the 9 inventory slots always displayed on the screen) and the rest of the player's main inventory. Representing this as a basic Venn Diagram we see:

![](http://eq2.co.uk/pr/inv/playerinv.png)

When working directly with Minecraft, the two inner inventory arrays are available directly as public members of the `InventoryPlayer` class. However some additional functionality is available which supports both the *ViewModel*-esque contract required by Containers and also some additional convenience functions for working with only the hotbar slots. Some notable drawbacks with this mixing of concerns is that the following become apparent when working with the `PlayerInventory`:

* Some methods treat the slots within the inventory as a contiguous set, and treat the **Armour Inventory** as indices beyond the **Main Inventory** indices whereas other code does not, this presents an inconsistent external interface.
* Conversely, some methods only deal with the hotbar slots, and not the rest of the Inventory
* No distinction is made between any positions in the inventory, thus external code has to "know" the meaning of slot indices within the Inventory in advance. *(A problem not confined to the `InventoryPlayer` class)*

It should be noted that this scenario is far from unique amongst inventories, and only gets worse when dealing with Containers. For example let's take a look at a container which you're likely to be very familiar with, `ContainerPlayer`:

![](http://eq2.co.uk/pr/inv/playercontnr.png)

This container presents a View of 3 separate Inventories (essentially ViewModels) which in turn represent 4 distinct underlying arrays of slots (essentially the underlying Models).

You may find yourself asking *"so what the hell does all this have to do with queries?"* and that's a fair question. Ultimately we have 3 ways of dealing with this mess in order to improve it from the point of view of API consumers:

1. Abstract the hell out of things. Provide "sensible" external interfaces and break everything down into small logical units which hide the underlying horribleness.
 * **Pros:**
     * easy to work with from a consumer's point of view
     * hides all the horrible internals
 * **Cons:**
     * nasty to maintain (if anything in the implementation changes then the abstraction could get more and more complex)
     * hard to extend (mod inventories and and kind of custom inventory are hard to work into this model because the abstractions themselves are "hard coded")
     * makes simple operations more complex than they need to be (consumers need to explicitly dig through layers of abstraction to get to what they want)
     * doesn't actually **add** any value, just hides all the nastiness
2. Expose the underlying horribleness but provide a metadata system which assigns meaning to the exposed data structures, such as allowing a "range" of slots to be described.
 * **Pros:**
     * doesn't add much overhead
     * keeps people who are used to the old system happy
     * provides added value in terms of some meta information about the meanings of slots
 * **Cons:**
     * just as brittle as the underlying impementation
     * doesn't provide any useful abstraction
3. *(Using queries)* Make metadata an intrinsic part of the API and allow consumers to stipulate exactly what part of an inventory they *mean* by accessing inventory *via the metadata using queries*:
 * **Pros:**
     * All the advantages of both option **1.** and **2.** (the underlying representations without metadata are *still accessible* via casting down - see implementation details below)
     * Consumers only ask for what they need, they never need to care about what *type* of Inventory or Container they're dealing with, only the results of a query
     * Intrinsically extensible - the nature of queries makes it easy for third-party extensibility without removing the ability to do simple `instanceof`-and-cast operations which would have been *required* before.
 * **Cons:**
     * May be difficult for consumers to adjust to the new model (although the nature of the implementation means that stick-in-the-muds can still use the old model if they're foolhardy enough)
     * Underlying implementation is more complex than other systems *(however it should be noted that all the complex logic will be handled in the initial implementation, and that ongoing __maintenance__ of the system will actually be __simpler__)*

Presented with the options it's clear that option 3, using queries has merit.

#### Goals for a Query-based implementation

Fundamental to the concept of querying Inventories is the basic premise that:

* An **Inventory** is a **View** of **one or more "sub inventories"**
* An **Inventory** and its **sub inventories** exist in a **parent-child** relationship
* An **Inventory** is thus a **View of Views** with arbitrary depth

In the `InventoryPlayer` example above, the relationship of the Inventory to its notional sub-inventories is as follows:

![](http://eq2.co.uk/pr/inv/invhierarchy.png)

The main idea of queries is that given an unknown `Inventory` instance, it should be possible to query for any sub-inventory or combination of matching sub-inventories, with the query returning either all sub-inventories which match the query, or an empty set if no sub-inventories matched the query.

It's useful to bear in mind that our above definition of `Inventory` essentially means that everything can be represented as an `Inventory` right down to `Slot` (defined as an `Inventory` with only a single position) and thus a more helpful representation when considering what can be returned by a query is the following:

![](http://eq2.co.uk/pr/inv/invhierarchydeep.png)

We're now in a position to specify what a **query** should actually return:

* A **Query** should return all **Sub Inventories** of the **Inventory** which match the supplied **criteria**
* If a **sub inventory** is **included** in the results set, its **parent** will **not be**
* A query will **never return a null** result, it will return an **Empty Inventory** object.
* A query will **never return duplicate** entries

Which produces as a consequence some assumptions for `Inventory` itself:

* The **result** of a query is always an `Inventory`
* An `Inventory` is `Iterable<Inventory>` and the returned iterator traverses the child nodes of the Inventory's hierarchy
* An `Inventory`'s leaf nodes should also be traversable via a method which returns an `Iterator`

To give some examples, based on the hierarchy above:

* A query for an imaginary `Hotbar.class` should return the Hotbar inventory
* A query for `ItemTypes.TNT` should return an Inventory with all of the slots containing TNT
* A query for the imaginary `InventoryRow.class` should return each row in the Main Inventory and the Hotbar

Which produces the following assumptions:

* When performing a query, if an `Inventory` matches the query directly, it should return itself
* When performing a query, an `Inventory` should perform a depth-first search of its hierarchy. A matching child node will remove its parent from the results set if present.
* If no children match the query, the query should return an `EmptyInventory`.
* If all direct children match the query, the query should return the parent *(eg. if all InventoryRow children within an InventoryGrid match the query, then the InventoryGrid is returned)*

We now have enough information to begin formulating our query interface.

### Inventory methods for query results

Since the result of a Query will always be an `Inventory`, we add some decoration and methods to our `Inventory` interface. Firstly, we have the `Inventory` extend `Iterable<Inventory>` as planned. Next we add a method for iterating the leaf nodes of our `Inventory` and a convenience function for checking whether the `Inventory` has no slots:

```java
interface Inventory extends Iterable<Inventory> {
	//
	// ... code code code (see above) ...
	//

	/**
	 * Returns an iterable view of all slots, use type specifier to
	 * allow easy pseudo-duck-typing
	 */
	public abstract <T extends Inventory> Iterable<T> slots();

	/**
	 * Returns true if this Inventory contains no children
	 */
	public abstract boolean isEmpty();
}
```

Depending on requirement, we **may** also wish to add some convenience methods to work with `Inventory`s which are result sets, for example

```java
	/**
	 * Return the first child inventory, effectively the same as
	 * Inventory::iterator().next() but more convenient when we are
	 * expecting a result set with only a single entry. Also use type
	 * specifier to allow easy pseudo-duck-typing. If no children, then
	 * returns this.
	 */
	public abstract <T extends Inventory> T first();

	/**
	 * Return the next sibling inventory, allows traversing the inventory
	 * hierarchy without using an iterator. If no more children, returns
	 * an EmptyInventory.
	 */
	public abstract <T extends Inventory> T next();
```


### Inventory methods for executing queries

It is anticipated that the full scope and possibilities of queries will only become apparent in time, and also that the initial system should be sufficiently extensible to support quite a wide scope of potential queries. However the basic query types which should be in the initial release are:

#### Query by type

```java
public abstract <T extends Inventory> T query(Class<?>... types);
```

Query for sub-inventories matching the specified interfaces/concrete classes (effictively an `instanceof` check). Multiple classes can be specified and logical `OR` will be applied.

**Example 1:**
```java
Inventory inv = ...; // An unknown inventory
PlayerInventory pinv = inv.query(PlayerInventory.class).first();
```

**Example 2: Logical AND** _(querying the result of a query)_
```java
Inventory inv = ...; // An unknown inventory
HotBar hotbar = inv
	.query(InventoryRow.class)
	.query(HotBar.class).first();
```

#### Query by contents

```java
public abstract <T extends Inventory> T query(ItemTypes... types);
```

Query for slots containing `ItemStack`s with items of the specified type. A logical `OR` applied between query operands.

**Example:**
```java
Inventory inv = ...; // An unknown inventory
Inventory slotsWithTNT = inv.query(ItemTypes.TNT);
if (!slotsWithTNT.isEmpty()) {
    // found slots with TNT!
}
```

#### Query by property

As well as defining a (possibly multi-dimensional) hierarchy of sub-inventories, it is anticipated that some sub-inventory types will tag their members with additional data which can be used to include them in queries. For example, let's assume that we have a `GridInventory` interface:

```java
/**
 * MetaInventory(?) stores arbitrary properties for child inventories
 */
interface MetaInventory extends Inventory {
	/**
	 * Get a property defined in THIS inventory for the specified
	 * (immediate) sub-inventory.
	 */
	public abstract <T extends InventoryProperty> getProperty(Inventory child, Class<T> property);
}

/**
 * This type of inventory arranges its children in a grid.
 */
interface GridInventory extends MetaInventory {
	/**
	 * Get the number of rows in the grid
	 */
	public int getRows();

	/**
	 * Get the number of columns in the grid
	 */
	public int getColumns();

	/**
	 * Get the X/Y position of the specified slot. The InventoryPos
	 * class extends InventoryProperty and thus this method is just a
	 * convenience method which effectively calls:
	 * 
	 *    this.getProperty(slot, InventoryPos.class);
	 */ 
	public abstract InventoryPos getSlotPos(Inventory slot);
}
```

Assuming that `InventoryPos` extends some ficticious base class `InventoryProperty`, we now have a mechanism for querying by any arbitrary properties we care to define:

```java
public abstract <T extends Inventory> T query(InventoryProperty... props);
```

Query for sub-inventories where the specified property is set and `.equals()` the supplied properties. Logical `OR` is applied between operands.

**Example:**
```java
Inventory inv = ...; // An unknown inventory
Inventory slots = inv.query(new InventoryPos(2, 2)).first();
```

#### Query by name
Since in the real world, `Inventory` classes are always nameable, we can use a `String` overload to query by name:

```java
public abstract <T extends Inventory> T query(String... names);
```

#### Query by arbitrary operands

To promote extensibilty, even where completely unknown Inventories are in play, we should also include a general query interface which will allow arbitrary queries to be executed:

```java
public abstract <T extends Inventory> T query(Object... args);
```

Query for sub-inventories using an arbitrary check defined by the inventory class in question.

**Example:**
```java
Inventory inv = ...; // An unknown inventory
SomeObject someObject = ...; // Something
Inventory result = inv.query(someObject); // ???
```

### Multi-Dimensional Query Model

So far, for simplicity we have outlined a straightforward hierarchical model of the inventory. However it is not the intention that underlying implementations should adopt such a simple model. To facilitate querying by dimensions, we can represent underlying sub-inventories in as many different dimensions as we like, and allow queries to return corresponding views. For example we can expand our simple example above to allow the Inventory to be queried by column:

![](http://eq2.co.uk/pr/inv/invhierarchymulti.png)

In this example the `InventoryGrid` contains both row and column sub-Inventories. This is by no means the limit of what can be represented by further dimensions. However note that one tree should always be considered the *master* view (a notional appellation), a traversal of which visits each leaf node in a deterministic order which is the iteration order as experienced by `slots()`.

Essentially every Inventory should have a deterministic *spanning tree*, whose depth-first traversal will be the primary iteration order of the leaf nodes. This order is left to implementors to decide upon. Here is an example for the inventory structure shown above.

**Example Spanning Tree for the Player Inventory**
![](http://eq2.co.uk/pr/inv/invhierarchyspanning.png)

**Example traversal of the Spanning Tree**
![](http://eq2.co.uk/pr/inv/invhierarchytraverse.png)

### Implications for Intended Usage

So far we have seen how queries can work in general, the main goals of incorporating queries into the API are to change the general way that consumers interact with Inventories. Without queries, a typical interaction with an Inventory might take place as follows:

1. Consumer obtains an Inventory instance from an API object. The inventory is of a known type - eg. a `PlayerInventory` is obtained from a `Player` because it is advertised by the object.
2. Knowing the type of inventory, the consumer performs some operations on the inventory: for example moving items from the main inventory to the hotbar, or consuming items from the inventory.

Now if the same code wishes to interact with a different type of Inventory, the code has to be adapted or rewritten, even if the other inventory has substantially the same behaviour or characteristics.

With a query-based approach, the interaction takes place as follows:

1. Consumer obtains an Inventory instance from an API object. The inventory is not a specialised type, the return type of the getter is simply `Inventory`.
2. The consumer queries for the inventory characteristics it requires, for example it queries for all `InventoryRow`s
3. The consumer interacts with the returned rows, or can simply skip processing if the query returns an empty set.

If the same code wishes to interact with a different type of Inventory, it can do so as long as the Inventory has `InventoryRow` children.

> **SIDE NOTE**
> Consider porting of old code. Old code may simply (as a stop-gap) wish to obtain the `PlayerInventory` and manipulate it directly, without any queries. With queries this is still supported as long as `PlayerInventory` is provided as an interface in the API. Consumers wishing to perform legacy interactions can either:
> 
> 1. Query directly for the `InventoryPlayer` interface *(this works because of assumption 1 of queries - if an inventory matches a query it simply returns itself)*
> 2. Use an `instanceof` check followed by a type cast
>
>This provides a convenient way for legacy code to be ported quickly to the new system, and transition to a full query-based implementation at the convenience of the plugin author.

Taking the intended usage into account, we can deduce some additional characteristics our API should have:

* Any API object which has an Inventory should simply return `Inventory`, with no requirement to return a particular subinterface

* Consumers should *always* obtain `Inventory` subclasses by querying for them (either directly or via `instanceof` check)

* Specific Inventory subinterfaces should still be provided within the API in order to satisfy queries, however they should not generally be returned and different implementations are therefore at liberty to select whichever interfaces they wish to implement - in other words there is no hard rule that says a `PlayerInventory` must have `InventoryRow`s as children for example, the query model means that this is purely optional and the only negative effect would be that consumers expecting such a structure would not be able to perform their operations.

 This works in both directions however, and while it is expected that all implementations will likely follow similar conventions (in order that queries can be relied upon across platforms), it's also possible that some implementations will be able to innovate, and extend their capabilities without in any way breaking backward compatibility.

 The implication of all this is that the shape and design of the underlying Inventories is thus free to follow a logical representation of the underlying game *without* at any point breaking backward compatibility.
 
* This in turn leads us to the conclusion that the number of "specific" inventories should be kept to a minimum, and general-purpose Inventory interfaces should be used wherever possible. The general-purpose Inventory interfaces should essentially represent *characteristics* of inventories, with little regard for specific Inventory types.

### Relationship with Containers

It should be clear by now that a query-based model also solves the Container problem. By acknowledging that the line between Inventory and Container can often be blurry, we can treat Containers as Inventories for the most part (since we have essentially turned Inventory into a ViewModel at this point anyway, and all a Container really is is a ViewModel) and simply have them exist as yet another type of SubInventory interface.

## Features *not* included within this PR

Taking a longer view, the following features are worthy of consideration

* **Inventory Transactions** - basically SQL transactions but for Inventories allowing *commit* and *rollback* operations

* **Aggregate Inventory Operations** - basically SQL join for queries, such that a logical union of inventories can be created and subsequently queried and operated upon as if it were a single inventory. *This isn't too hard to do I just don't have sufficient time to flesh it out*.

* **Atomic Multi-Inventory Operations** - as a poor second choice to full-blown transactional behaviour, having atomic operations (get-stack-from-first-inventory, deposit-in-second) would still be useful.

* **Cookies** - congratulations, if you read this far you get a cookie *(Not a real cookie, an imaginary one. What? You think I'm made of cookies or something?)*

gabizou commented:
> Cookies - congratulations, if you read this far you get a cookie (Not a real cookie, an imaginary one. What? You think I'm made of cookies or something?)

Can I have a cookie for being patient and waiting for this PR to be made?

Mumfrey commented:
@gabizou no you may not.

gratimax commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -0,0 +1,230 @@
+ * TODO Flesh out javadoc from proposal document. For now, see proposal doc
+ * here: https://github.com/SpongePowered/SpongeAPI/pull/443
+ */
+public interface Inventory extends Iterable<Inventory>, Nameable {
```
Do all inventories really have Translatable names, even player-created ones?

Mumfrey replied:
I considered putting this on `Container` instead but in vanilla minecraft every `IInventory` is `IWorldNameable` and thus anything which is an Inventory basically has a name. I extracted that out to `Nameable` partly because it's a useful interface to have on its own merits but also because we could potentially graft the `Nameable` onto the hierarchy somewhere else if we don't like it being on the base interface.

Food for thought at any rate. Where I'm not sure about stufff I've erred on the side of simplicity and consistency with the view that we can throw it all in the bin if we feel the inclination.

Zidane replied:
Wouldn't the name of the inventory be a Translatable?

gabizou commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -536,4 +539,20 @@
     List<BannerPatternShape> getBannerPatternShapes();
 
+    /**
+     * Retrieves the ItemDictionary for this GameRegistry
```
You call it ItemDictionary but it returns GameDictionary?

Mumfrey replied:
Don't look at me, I didn't write that bit.

gratimax replied:
I take the blame. I did not expect the original PR to be merged or used at all, so I didn't perfect it. There are still a few rough spots that need some smoothing and thought. I can fix those parts if needed.

Mumfrey replied:
Hey don't worry about it, I didn't do much more than look over the code I didn't alter, I was anticipating as you were that stuff would fall out of the review process anyway and that any dumb mistakes, glaring omissions or just minor cruft would be caught in the filter.

Minecrell commented:
Some small notes to the code style:
- We don't use any redundant modifiers on interface methods to keep the interfaces clean, so we need to remove the `public abstract` prefixes in _some_ of the interfaces here.
- Some classes are missing Javadocs and some methods are missing `@param` and `@return` tags.
- The normal Javadoc description should have an period at the end, the description of the tags shouldn't.

I've also noticed a few events in here that are already in the SpongeAPI.

Minecrell commented on src/main/java/org/spongepowered/api/event/entity/EntityEquipEvent.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Raised when an entity equips an item.
+ */
+public interface EntityEquipEvent extends EntityEvent, InventoryEvent {
```
Is this the same as [`EntityEquipmentChangeEvent`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/event/entity/EntityEquipmentChangeEvent.java)?

Minecrell commented on src/main/java/org/spongepowered/api/event/inventory/FurnaceBurnEvent.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Fired when an ItemStack is burned in a furnace as fuel.
+ */
+public interface FurnaceBurnEvent extends BlockInventoryEvent, Cancellable {
```
Is this the same as [`FurnaceConsumeFuelEvent`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/event/block/data/FurnaceConsumeFuelEvent.java)?

Minecrell commented on src/main/java/org/spongepowered/api/event/inventory/FurnaceSmeltEvent.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Fired when a furnace smelts a source ItemStack into a resultant ItemStack.
+ */
+public interface FurnaceSmeltEvent extends BlockInventoryEvent, ItemResultEvent, Cancellable {
```
[`FurnaceSmeltItemEvent`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/event/block/data/FurnaceSmeltItemEvent.java)

Minecrell commented on src/main/java/org/spongepowered/api/item/inventory/custom/CustomInventoryBuilder.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @param name The title
+     */
+    CustomInventoryBuilder setName(Translatable name);
```
This should be just `name()` with no prefix

Minecrell commented on src/main/java/org/spongepowered/api/item/recipe/ShapedRecipeBuilder.java:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @param width The width of the grid
+     */
+    void setWidth(int width);
```
Same here, no prefix

Mumfrey commented:
>     We don't use any redundant modifiers on interface methods to keep the interfaces clean, so we need to remove the public abstract prefixes in *some* of the interfaces here.

Okay that's fair enough, this is my first contribution to the API so I expected their would be style differences. I always use the modifiers just because it seems more readable but I can remove them no problem.

>     Some classes are missing Javadocs and some methods are missing @param and @return tags.

To be honest the plan was to flesh out the javadoc later, I wanted to wait until all the I's were dotted and T's were crossed but I've been too busy and @Zidane was probably going to murder me if I took any longer. Basically take a view that if there *is* some Javadoc, it'll get fixed later when I have two seconds to rub together, either that or someone else can beat me to it.

>     The normal Javadoc description should have an period at the end, the description of the tags shouldn't.

This isn't anywhere in the style guide, I'm not sure how I was supposed to know this.

> I've also noticed a few events in here that are already in the SpongeAPI.

Likely because @gratimax already added them before they were added to the API. I'm sure redundant events can be removed. Any ideas which ones?

Minecrell commented:
@Mumfrey Wasn't sure if the PR was already ready to be reviewed on code style, so I just thought I do that so we can fix them. :)

> This isn't anywhere in the style guide, I'm not sure how I was supposed to know this.

Some more things are mentioned in our [contribution guidelines](https://docs.spongepowered.org/en/latest/devs/guidelines/#code-style) although it's not complete. I guess you were not able to know this, but now you do. ;)

> Likely because @gratimax already added them before they were added to the API. I'm sure redundant events can be removed. Any ideas which ones?

I've commented on the (possibly) duplicate events above.

Mumfrey commented:
> @Mumfrey Wasn't sure if the PR was already ready to be reviewed on code style, so I just thought I do that so we can fix them. :)

Personally I tend to focus on the important aspects of code style (readability, wrapping, stuff having *some* kind of javadoc rather than none) and deal with the more fiddly parts in a second pass, at this point as long as it builds then people can review the structure of the PR and other issues can be mopped up going forward.

> Some more things are mentioned in our contribution guidelines although it's not complete. I guess you were not able to know this, but now you do. ;)

I was including the contribution guidelines in my above assertion, there is no mention of full stops other than specifically mentioning they should be omitted on *at clauses*. 

> I've commented on the (possibly) duplicate events above.

I'll leave that one for @gratimax to take a view on. I've removed the `public abstract` modifiers and added javadoc to two classes which were missing it. The rest can wait for a second pass because there are more pressing things that need work.

ZephireNZ commented:
First off, glorious pull request.

How, in this new system, would you go about taking X amount of items from a given inventory? I imagine you would you create a query for the ItemType, but then what? Iterate over that inventory, removing item stacks until you've removed enough?

Mumfrey commented:
> First off, glorious pull request.
> 
> How, in this new system, would you go about taking X amount of items from a given inventory? I imagine you would you create a query for the ItemType, but then what? Iterate over that inventory, removing item stacks until you've removed enough?

Yes, because remember that your result set is an `Inventory`. So basically if you want to consume a specific item from anywhere in the inventory you query for slots containing the inventory and then `peek` to get the first stack. Although now you ask it's given me an idea that maybe you could instead `poll` with a `limit` argument to allow you to consume items up to a set amount.

You need not iterate over the inventory because you can consume items directly from the query result (since it's an `Inventory`) so basically to consume a single item (eg to consume a arrow when you fire it).

```java
Inventory arrows = player.getInventory().query(ItemTypes.ARROW);
if (arrows.isEmpty()) {
    // boo, no arrows
    return;
}

ItemStack quiver = arrows.peek().get(); // We know it's present so calling get() is safe
quiver.setQuantity(quiver.getQuantity() - 1);
```

However, I think a nicer way (which I will add now you mention it) would be:

```java
Inventory arrows = player.getInventory().query(ItemTypes.ARROW);
if (arrows.isEmpty()) {
    // boo, no arrows
    return;
}

ItemStack arrow = arrows.poll(1).get(); // Consume a stack of 1 array from the result
```

modwizcode commented:
This is awesome in all of its definitions, what led you to come up with a system so different?

Mumfrey commented:
@modwizcode there's honestly nothing really outlandish in this proposal, it's more the fact that everyone is so entrenched in the existing Inventory architecture (which, let's face it, is bloody awful) that nobody has really taken a step back and said "wait a minute, how **should** this work?" So much interaction with inventories to date has required far to much prior knowledge of the internals of the inventory in question, I simply tried to come up with a more expressive way of interacting with inventories.

kitskub commented:
Quick comment because I'm on my phone and can't really look through the code: instead of `new InventoryPos(x, y)`, I think `InventoryPos.of(x, y)` would be cleaner.

Mumfrey commented:
@kitskub they're just the interfaces, how they're implemented is up to the implementation.

kitskub commented:
@Mumfrey okay. Sorry, like I said: can't read the code at the moment.

Mumfrey commented:
@kitskub It might be nice to have some concrete implementations in the API but I'm not sure what format that kind of thing is taking and it's up to @Zidane et al. to figure out the best way to factory these objects that's consistent with how things are done elsewhere in the API.

boformer commented:
Well done. 2 things:

* Also add a ``peek(int limit)`` method
* Add a way to count the number of single items contained in the inventory

And a simple question:
How can I place an item stack (e.g. a server rule book) into the first slot of the hotbar, even if the slot is filled with an existing item?

My Solution:
First, get the ``Inventory`` of the slot:  ``playerInv.query(Hotbar.class).query(new InventoryPos(0,0)``

Then ``offer`` the book item stack to the slot. If it was rejected, ``poll`` the existing item stack from the hotbar slot.

Then ``offer`` the book item stack to the hotbar slot again.

Then ``offer`` the replaced item stack to the ``playerInv``.

If it was rejected. revert everything so that no items are lost (and drop the book).

Does it work like this?

Mumfrey commented:
> Also add a peek(int limit) method

Hmm. Interesting. What's the use-case just for my curiosity?

> Add a way to count the number of single items contained in the inventory

Okay.

> And a simple question:
> How can I place an item stack (e.g. a server rule book) into the first slot of the hotbar, even if the slot is filled with an existing item?

```java
ItemStack rulebook = ...;
inventory.query(Hotbar.class).first().set(rulebook);
```

for other slots:

```java
Inventory result = inventory.query(Hotbar.class);
if (!result.isEmpty()) { // check the query actually succeeded
    Hotbar hotbar = result.<Hotbar>first();
    hotbar.set(3, rulebook); // Hotbar is an OrderedInventory so we can use positional set
}```

or using a fully query-based version:

```java
inventory.query(Hotbar.class).query(new SlotIndex(3)).set(rulebook);  
```

ST-DDT commented:
First of all: Awesome API! (And nice and complete description with descriptive images)

@Mumfrey I guess @boformer tried to add an `ItemStack` to the `Hotbar` Slot 0. And moves any existing Item at that slot to the normal player inventory. (including later slots in the hotbar?


I have three question myself. 
* in one of the pictures you displayed above you specified the iteration order to iterate over the `InventoryRows`, but it is possible to iterate over `InventoryColumns` as well? 
* If the `Inventory` that is queried is a PlayerContainer as you called it (with the crafting slots), will they be included in the iteration as well?
* How do i set the active hotbar index/slot?

kenzierocks commented:
@ST-DDT I assume it would be like this:

```java
inventory.query(Hotbar.class).query(new SlotIndex(/* selected slot here */)).set(rulebook);  
```

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -0,0 +1,263 @@
+     * The maximum number of stacks the Inventory can hold. Always 1 for
+     * {@link Slot}s and always 0 for {@link EmptyInventory}s.
+     */
+    int capacity();
```
A method for summing up all `ItemStack` sizes would be nice.
`Inventory.query(ItemTypes.ARROW).totalCount();`
Returns the total number of arrows in the inventory.

Mumfrey replied:
Already suggested below.

ST-DDT commented:
@kenzierocks IMO this will replace the item at the given slot, removing it from the inventory, instead of moving it to another slot. (This is what i expect the method to do)

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -0,0 +1,263 @@
+     * objects in the inventory as required to accomodate the entire stack. The
+     * entire stack is always consumed.
+     */
+    void set(ItemStack stack);
```
Maybe this should return `Optional<ItemStack>` for cases where an other Item has been thrown out of the inventory in order to add said item. (`Map.put()` also returns any previous entry)

Mumfrey replied:
That's better handled by an event imo.

ST-DDT replied:
> has been thrown out of the inventory
The old item in the slot that has been replaced by the new item.

Using events to get the replaced item would totally break the call chain here and will most likely force me to never run this method in a situation where it replaces another item.

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -0,0 +1,263 @@
+     * against the child inventory passing the unknown operand as an argument.
+     * </p>
+     */
+    <T extends Inventory> T query(Object... args);
```
Maybe some methods to query using `Predicate`s would be nice.
`query(Predicate<Slot> contidtion)`
For example you query for `GOLDBLOCK`s but also allow `ItemStack`s where the Item's Name is `coin`.

````java
new Predicate() {

    public boolean apply(Slot input) {
        final ItemStack item = input.peek();
        if (item.getType() == ItemTypes.GOLDBLOCK) {
            return true;
        } else {
            return "coin".equalsIgnoreCase(item.getItemName());
        }
    }

}
````

Mumfrey replied:
No, because `Predicate` indicates that the logic is supplied by the user. The comparison logic is implicit in the `InventoryProperty` passed in and the logic is deterministically implemented by the inventory itself.

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -0,0 +1,263 @@
+    /**
+     * Get without removing the first available stack from this Inventory
+     */
+    Optional<ItemStack> peek();
```
You should mention whether this returns a copy or a life item. Otherwise the following code's result is undefined.
````java
inventory.query(ItemTypes.ARROW).peek().setQuantity(64);
````

Zidane replied:
ItemStacks are inherently mutable so I'd expect this to return the live ItemStack.

Good question, this should be defined.

Deamon5550 replied:
Since no where in this class or the documentation is any mention of object copies being thrown around I'd say its safe to assume that objects are live unless specified otherwise and needing to specify it on every method is rather overkill. Most people are familiar with the concept of a Stack so these methods shouldn't be alien to them.

ST-DDT replied:
@Deamon5550 I'm from Bukkit so i had mutable `ItemStack`s as well, but they never haven been live (proxies). So i had to set them after modifying them.

Mumfrey replied:
@Zidane correct, this is specified with the intention that the returned stack is "live".

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/ItemStackComparators.java:
```diff
@@ -37,56 +36,69 @@
-     * Compares ItemStacks based on {@link ItemType}. This comparator will not
-     * return the same results as ItemStack.equals(ItemStack) for ItemStacks
-     * with extra attached data, different damage values, or different sizes.
+     * Compares ItemStacks based on {@link org.spongepowered.api.item.ItemType}.
```
For readability reasons you should import `ItemType` instead of fully qualify it all the time.

Mumfrey replied:
Except that generates a warning since the import is "unused", so this way is better when there's only a single usage.

ST-DDT replied:
`ItemType` and `ItemStack` are both used more than once.
In addition to that `ItemStack` is also used as return type.

I haven't seen any warnings from gradle nor my IDE here.

simon816 replied:
They're not "unused" since it's being used in the javadoc. That warning is wrong

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/ItemStackComparators.java:
```diff
@@ -37,56 +36,69 @@
     private ItemStackComparators() {
     }
 
-    private static final class Type implements Comparator<ItemStack> {
```
Is there a reason why you removed the `private` modifier here?

Mumfrey replied:
Because private inner classes are stupid and pointless.

ST-DDT replied:
They are private (and final), because they really are anonymous classes, that have been refactored to inner classes to improve readability in this util class and are not meant to be instantiated more than once.

Due to its location in the API and its intended use, there is simply no point of making it anything else than private.
(It does not matter, but from my point of view, there is no benefit from removing the private modifier, because its intended as anonymous class.)

simon816 replied:
Should really be `private` since it's used only internally in this class. Plugins should not have direct access to these classes, but should use the static fields instead.

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/armour/ArmourInventory.java:
```diff
@@ -0,0 +1,85 @@
+ * Armour inventory for {@link org.spongepowered.api.item.inventory.Carrier}s
+ * that can wear armour.
+ */
+public interface ArmourInventory extends OrderedInventory, CarriedInventory<ArmorEquipable> {
```
Maybe this can be renamed to `EquipementInventory` to allow this be used for horses as well.
(Or for mod provided backpacks as well)

Mumfrey replied:
There's no reason it couldn't be, except that I'd probably *add* EquipmentInventory at this point and have ArmourInventory subclass it, because ArmourInventory has a specific purpose (slots can be queried by ArmourType.

ST-DDT replied:
I thought the other way round.
You could replace the ArmourType with EquipmentType to allow query for HorseArmor/Saddle/WeaponInHand....
That way this interface is not limited to `Player`s allone, but can also be used for Zombies, Skeletons, Horses and many more, including future mobs.

Mumfrey replied:
No it would definitely be the other way around, because an ArmourInventory would essentially be a specialised EquipmentInventory, and ArmourType is a specialised EquipmentType. Beacon Blocks are already catered for with the *fields* thing, crappy as that is.

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/types/GridInventory.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Get and remove the stack at the supplied position in this Inventory
+     */
+    Optional<ItemStack> get(int x, int y);
```
Maybe this should be called `poll()` to match the base inventory's method name.

Mumfrey replied:
Probably.

Zidane commented:
@Mumfrey 

How much more do you want to change of this before I can pull this into master?

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/types/OrderedInventory.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Get and remove the stack at the supplied index in this Inventory
+     */
+    Optional<ItemStack> get(int index);
```
Should be called `remove(int index)` to be similar to the `List`s `remove(int index)` method.

ST-DDT commented on src/main/java/org/spongepowered/api/item/inventory/types/OrderedInventory.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Set the item in the specified slot 
+     */
+    void set(int index, ItemStack stack);
```
This should probably return any old Item that may have been present at the given index.

ST-DDT commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -536,4 +539,20 @@
+     *
+     * @return The recipe registry
+     */
+    RecipeRegistry getRecipeRegistry();
```
Why is the `RecipeRegistry` separated from the `GameDictionary`

gratimax replied:
they're different things, look at the javadocs

Mumfrey commented:
@Zidane I'm going to change the specific properties into concrete classes because I've realised that having them as interfaces doesn't serve any purpose since they can still be extended by implementors and their behaviour is actually fixed.

Other than that all that needs to happen is @gratimax needs to take a view on the duplicate events.

progwml6 commented:
@mumfrey all names are using US English in the API

modwizcode commented:
Actually we use non US spellings for Canceled i.e: "Cancellable"

modwizcode commented:
@Mumfrey What is your plan for the transaction operations?

modwizcode commented:
This doesn't provide a way to add items to an inventory as far as I can tell. Unless I'm missing something, I cannot find a way to actually get an ItemStackBuilder.

heni123321 commented:
@modwizcode there is Game.getRegistry().getItemBuilder() that returns a  ItemStackBuilder

modwizcode commented:
Ah ok.

Aaron1011 commented:
A few suggestions:

* Clarify the behavior of `Inventory#getProperty` when multiple `InventoryProperty`s could be returned
  * This could occur if two subtypes of, for example, `IntPropery`, are sent, and `getProperty(IntProperty.class)` is called
* Allow multiple properties of the same class (or with a common ancestor) to be set. This is related to the above point
* Allow querying an Inventory with a specific `InventoryPropery` *class*, as opposed to a particular instance.

Mumfrey commented:
> * Clarify the behavior of Inventory#getProperty when multiple InventoryPropertys could be returned

An InventoryProperty is never returned, only used to match

> * This could occur if two subtypes of, for example, IntPropery, are sent, and getProperty(IntProperty.class) is called

That's not how matching with properties work. If there are multiple IntProperty then they *must* have different keys, and thus the keys for the supplied matching properties will be different.

> Allow multiple properties of the same class (or with a common ancestor) to be set. This is related to the above point

You can as long as they have different keys

> Allow querying an Inventory with a specific InventoryPropery class, as opposed to a particular instance.

This doesn't make sense, you need an *instance* for a query by property to be meaningful in any way, testing for the *presence* of a property doesn't have any meaning.

Aaron1011 commented:
> An InventoryProperty is never returned, only used to match
It's returned [here](https://github.com/SpongePowered/SpongeAPI/pull/443/files#diff-6465f1b99c02dc355955b64306e30232R204)

Mumfrey commented:
> It's returned here

Ooooh! I see what you mean. Yes, that method needs an override which accepts the key, well spotted.

Mumfrey commented:
@Aaron1011 [fixed](http://bit.ly/1DmX0BP)

Mumfrey commented:
@Zidane as far as I'm concerned this is ready to roll, if you get @gratimax to just take a look and remove/change those duplicated events then it's good to go.

simon816 commented on src/main/java/org/spongepowered/api/entity/hanging/Hanging.java:
```diff
@@ -43,7 +43,7 @@
      *
      * @param direction The direction to face
-     * @param forced Whether physics is ignored
+     * @param forced    Whether physics is ignored
```
Unnecessary extra spaces.

Mumfrey replied:
I think the world will keep turning somehow.

ST-DDT replied:
Maybe. But those extra formatting just blows up the entire PR. Although some of them may be useful.
(You/the team made the rules, it would be very nice to know those rules apply to Teammembers as well.)

IMHO @simon816 is right with this one.

Mumfrey replied:
The rules apply to us, but unlike PRs from non-team members, I'm free to do a cleanup pass on this later on when there aren't more important things that need to be handled first. Saying "oh well, let's keep this important PR on ice just because it's missing a few lines of javadoc" is stupid when it can be merged, people can get to work on things, then next week when I have time I can fix the formatting issues in one clean pass.

Asking a team member to do that on the third-party PR, or accepting two PRs for every PR (one for functionality and another to clean up) isn't realistic and it's much easier to ensure that PRs are "clean" before they're pulled. This is not the case here, I have more pressing stuff to be spending time working on, the javadoc and the little code style issues can wait.

ST-DDT replied:
Yeah, you are right. We just want to make sure nothing slips attention because it is less important.

simon816 commented on src/main/java/org/spongepowered/api/event/inventory/CraftItemEvent.java:
```diff
@@ -0,0 +1,69 @@
+public interface CraftItemEvent extends ViewerEvent, Cancellable {
+
+    /**
+     * Retrieves the CraftingInventory involved with this event.
```
Should `@link CraftingInventory` or write in plain English i.e "crafting inventory". (I prefer `@link`)
This applies to several other places too.

simon816 commented on src/main/java/org/spongepowered/api/event/inventory/ItemExtractEvent.java:
```diff
@@ -0,0 +1,28 @@
+ */
+package org.spongepowered.api.event.inventory;
+
+public class ItemExtractEvent {
```
This doesn't appear to be finished. Missing javadoc, no event hierarchy, should be an `interface`.

simon816 commented on src/main/java/org/spongepowered/api/item/ItemType.java:
```diff
@@ -44,8 +43,8 @@
-     * Get the default maximum quantity for {@link ItemStack}s
-     * of this item.
+     * Get the default maximum quantity for
+     * {@link org.spongepowered.api.item.inventory.ItemStack}s of this item.
```
Fully qualifying doesn't look or read as well compared to using a shorthand w/ import

simon816 commented on src/main/java/org/spongepowered/api/item/inventory/Carrier.java:
```diff
@@ -0,0 +1,41 @@
+import org.spongepowered.api.item.inventory.types.CarriedInventory;
+
+/**
+ * A Carrier is simply something that carries an {@link Inventory}
```
Missing full-stop.

simon816 commented:
I think there should be a list of standard keys for `GameDictionary` since there may be issues where a plugin registers something under "wood" but another plugin uses "planks".
`GameDictionary` only supports items, should it be called `ItemDictionary` instead? Or what about extending it's functionality to supporting `BlockType`s an possibly other game objects too.

simon816 commented on src/main/java/org/spongepowered/api/item/inventory/EmptyInventory.java:
```diff
@@ -0,0 +1,34 @@
+ * An EmptyInventory is an Inventory with exactly zero slots, no children and
+ * a guarantee that all queries will only every return an EmptyInventory.
+ */
+public interface EmptyInventory extends Inventory {
```
With such a fixed definition of what this Inventory must do, could this be implemented as a `public final class`?

Mumfrey replied:
No, because the semantics of this are important and need to be handled by an implementation.

simon816 commented on src/main/java/org/spongepowered/api/item/inventory/equipment/EquipmentInventory.java:
```diff
@@ -0,0 +1,88 @@
+     * Get and remove the stack for the specified equipment type in this
+     * Inventory
+     */
+    Optional<ItemStack> get(EquipmentSlotType equipmentType);
```
`pop` may be a better name for a method that gets and removes a stack.

Mumfrey replied:
it's supposed to be `poll` actually, I must have missed one. `WILLFIX`

simon816 commented on src/main/java/org/spongepowered/api/item/inventory/types/GridInventory.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Get and remove the stack at the supplied position in this Inventory
+     */
+    Optional<ItemStack> poll(int x, int y);
```
Why not `Vector2i`?

Mumfrey replied:
Because you can already use `SlotPos`.

simon816 commented on src/main/java/org/spongepowered/api/item/inventory/types/GridInventory.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Gets the number of columns in the inventory.
+     *
+     * @return The width of this ItemGrid.
```
Looks like `ItemGrid` as renamed to `GridInventory` but the javadoc got left behind.
Doc is also outdated in `item.recipe.Recipe` too

simon816 commented on src/main/java/org/spongepowered/api/item/recipe/RecipeRegistry.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @param recipe The Recipe to unregister
+     */
+    void remove(Recipe recipe);
```
Return `boolean` whether it was removed? (recipe may not have existed in the registry)

simon816 commented on src/main/java/org/spongepowered/api/item/recipe/ShapelessRecipe.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @return The ingredients
+     */
+    List<ItemStack> getIngredients();
```
What about a `Collection` instead since it doesn't have a shape does that imply it doesn't have an order?

simon816 commented on src/main/java/org/spongepowered/api/util/inventory/Coerce.java:
```diff
@@ -0,0 +1,417 @@
+     */
+    public static String toString(@Nullable Object obj) {
+        if (obj == null) {
+            return ""; 
```
Returning "null" (the string) may be more useful.

Mumfrey replied:
There is an underlying method to the madness of `Coerce`. The appropriate return here is an empty string.

ST-DDT commented:
@simon816 If i got the concept of the `GameDictionary` right, it will contain all stuff that can be registered by plugins. Achievements, Statistics, BlockTypes, ItemTypes, (EntityTypes), ...
Otherwise the Achievement PR would not be pending due to this PR.

ST-DDT commented on src/main/java/org/spongepowered/api/util/inventory/Coerce.java:
```diff
@@ -0,0 +1,417 @@
+        }
+        return list;
+    }
+
```
private Constructor missing

Mumfrey replied:
Ok, will add one.

Mumfrey replied:
4c42372bd4e6e9cba51ded10f9b246f9d69bf744

me4502 commented:
Is this PR mergeable?

olivervscreeper commented:
So how would you backup an entirw inventory with this system? Oh, and could you make inventories serializable by default?
